File: android/app/src/main/java/com/seyedjafari/prayertimes/PrayerCore.kt
package com.seyedjafari.prayertimes

import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * Kotlin port of shared/prayer-core.js — kept in sync with
 * shared/parsing-contract.md. Pure JVM (java.time only), no Android deps,
 * so it is unit-testable on the JVM.
 */
data class PrayerTimes(
  val city: String,
  val dawn: String,
  val sunrise: String,
  val noon: String,
  val maghrib: String,
  val midnight: String,
  val nextName: String,
  val nextTime: String,
  val stale: Boolean = false,
)

object PrayerCore {
  val CITIES = listOf("London", "Cardiff", "Glasgow", "Manchester", "Newcastle")
  const val SOURCE_BASE = "https://ic-el.uk/wp-content/icel/praying_timetable/prayer_times_en.php"
  val LONDON: ZoneId = ZoneId.of("Europe/London")

  private data class Row(
    val day: Int, val dawn: String, val sunrise: String, val noon: String,
    val maghrib: String, val midnight: String, val isToday: Boolean,
  )

  fun buildUrl(city: String, y: Int, m: Int): String = "$SOURCE_BASE?year=$y&city=$city&month=$m"

  private val rowRe = Regex("""<tr ID="(OddRows|EvenRows|TodayRow)">([\s\S]*?)</tr>""")
  private val cellRe = Regex("""<td[^>]*>([\s\S]*?)</td>""")
  private val tagRe = Regex("""<[^>]*>""")
  private val hhmm = Regex("""^\d{2}:\d{2}$""")

  private fun parseRows(html: String): List<Row> = rowRe.findAll(html).mapNotNull { mr ->
    val cells = cellRe.findAll(mr.groupValues[2])
      .map { tagRe.replace(it.groupValues[1], "").trim() }.toList()
    if (cells.size >= 7) Row(
      cells[1].toIntOrNull() ?: -1, cells[2], cells[3], cells[4], cells[5], cells[6],
      mr.groupValues[1] == "TodayRow",
    ) else null
  }.toList()

  private fun toMin(s: String): Int { val p = s.split(":"); return p[0].toInt() * 60 + p[1].toInt() }
  private fun fromMin(x: Int): String { val h = (x / 60) % 24; val m = x % 60; return "%02d:%02d".format(h, m) }
  private fun valid(s: String): Boolean = hhmm.matches(s) && s != "00:00"

  private fun repair(rows: List<Row>, idx: Int, get: (Row) -> String): String {
    val v = get(rows[idx]); if (valid(v)) return v
    var prev: String? = null; var next: String? = null
    for (i in idx - 1 downTo 0) if (valid(get(rows[i]))) { prev = get(rows[i]); break }
    for (i in idx + 1 until rows.size) if (valid(get(rows[i]))) { next = get(rows[i]); break }
    return when {
      prev != null && next != null -> fromMin(Math.round((toMin(prev) + toMin(next)) / 2.0).toInt())
      prev != null -> prev
      next != null -> next
      else -> v
    }
  }

  private fun computeNext(t: Map<String, String>, nowMin: Int): Pair<String, String> {
    val order = listOf(
      "Dawn" to "dawn", "Sunrise" to "sunrise", "Noon" to "noon",
      "Maghrib" to "maghrib", "Midnight" to "midnight",
    )
    val ev = order.map { (name, key) ->
      var mm = toMin(t[key]!!); if (key == "midnight" && mm < 180) mm += 1440
      Triple(name, t[key]!!, mm)
    }
    val up = ev.filter { it.third > nowMin }.sortedBy { it.third }
    val n = if (up.isNotEmpty()) up[0] else ev[0]
    return n.first to n.second
  }

  fun extractToday(html: String, todayDay: Int, nowMin: Int, city: String): PrayerTimes {
    val rows = parseRows(html)
    require(rows.isNotEmpty()) { "no rows" }
    var idx = rows.indexOfFirst { it.isToday }
    if (idx < 0) idx = rows.indexOfFirst { it.day == todayDay }
    require(idx >= 0) { "today not found" }
    val row = rows[idx]
    val times = mapOf(
      "dawn" to repair(rows, idx) { it.dawn },
      "sunrise" to repair(rows, idx) { it.sunrise },
      "noon" to repair(rows, idx) { it.noon },
      "maghrib" to row.maghrib,
      "midnight" to row.midnight,
    )
    val (nn, nt) = computeNext(times, nowMin)
    return PrayerTimes(
      city, times["dawn"]!!, times["sunrise"]!!, times["noon"]!!,
      times["maghrib"]!!, times["midnight"]!!, nn, nt,
    )
  }

  /** year, month, day in Europe/London. */
  fun nowLondon(): Triple<Int, Int, Int> {
    val z = ZonedDateTime.now(LONDON)
    return Triple(z.year, z.monthValue, z.dayOfMonth)
  }

  /** minutes-of-day in Europe/London. */
  fun nowMinLondon(): Int {
    val z = ZonedDateTime.now(LONDON)
    return z.hour * 60 + z.minute
  }
}

File: homey/lib/prayer-core.js
"use strict";
const CITIES = ["London", "Cardiff", "Glasgow", "Manchester", "Newcastle"];
const SOURCE_BASE = "https://ic-el.uk/wp-content/icel/praying_timetable/prayer_times_en.php";

function londonNow(date = new Date()) {
  const fmt = new Intl.DateTimeFormat("en-GB", {
    timeZone: "Europe/London", year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", minute: "2-digit", hour12: false,
  });
  const p = Object.fromEntries(fmt.formatToParts(date).map((x) => [x.type, x.value]));
  return { y: +p.year, m: +p.month, d: +p.day, minutes: (+p.hour % 24) * 60 + +p.minute };
}

function buildUrl(city, y, m) {
  return `${SOURCE_BASE}?year=${y}&city=${encodeURIComponent(city)}&month=${m}`;
}

function parseRows(html) {
  const rowRe = /<tr ID="(OddRows|EvenRows|TodayRow)">([\s\S]*?)<\/tr>/g;
  const cellRe = /<td[^>]*>([\s\S]*?)<\/td>/g;
  const rows = [];
  let m;
  while ((m = rowRe.exec(html))) {
    const isToday = m[1] === "TodayRow";
    const cells = [];
    let c;
    while ((c = cellRe.exec(m[2]))) cells.push(c[1].replace(/<[^>]*>/g, "").trim());
    if (cells.length >= 7) {
      rows.push({
        dow: cells[0], day: parseInt(cells[1], 10),
        dawn: cells[2], sunrise: cells[3], noon: cells[4],
        maghrib: cells[5], midnight: cells[6], isToday,
      });
    }
  }
  return rows;
}

const isValid = (s) => /^\d{2}:\d{2}$/.test(s);
const toMin = (s) => { const [h, m] = s.split(":").map(Number); return h * 60 + m; };
const fromMin = (x) => {
  const h = Math.floor(x / 60) % 24, m = x % 60;
  return String(h).padStart(2, "0") + ":" + String(m).padStart(2, "0");
};

function repairField(rows, idx, field) {
  const val = rows[idx][field];
  if (isValid(val) && val !== "00:00") return val;
  let prev = null, next = null;
  for (let i = idx - 1; i >= 0; i--) if (isValid(rows[i][field]) && rows[i][field] !== "00:00") { prev = rows[i][field]; break; }
  for (let i = idx + 1; i < rows.length; i++) if (isValid(rows[i][field]) && rows[i][field] !== "00:00") { next = rows[i][field]; break; }
  if (prev && next) return fromMin(Math.round((toMin(prev) + toMin(next)) / 2));
  return prev || next || val;
}

function computeNext(t, nowMin) {
  const order = [["Dawn", "dawn"], ["Sunrise", "sunrise"], ["Noon", "noon"], ["Maghrib", "maghrib"], ["Midnight", "midnight"]];
  const ev = order.map(([name, key]) => {
    let mm = toMin(t[key]);
    if (key === "midnight" && mm < 180) mm += 1440;
    return { name, time: t[key], min: mm };
  });
  const up = ev.filter((e) => e.min > nowMin).sort((a, b) => a.min - b.min);
  const n = up.length ? up[0] : ev[0];
  return { name: n.name, time: n.time };
}

function extractToday(html, L) {
  const rows = parseRows(html);
  if (!rows.length) throw new Error("no rows parsed");
  let idx = rows.findIndex((r) => r.isToday);
  if (idx < 0) idx = rows.findIndex((r) => r.day === L.d);
  if (idx < 0) throw new Error("today row not found");
  const row = rows[idx];
  const times = {
    dawn: repairField(rows, idx, "dawn"),
    sunrise: repairField(rows, idx, "sunrise"),
    noon: repairField(rows, idx, "noon"),
    maghrib: row.maghrib,
    midnight: row.midnight,
  };
  return { times, next: computeNext(times, L.minutes) };
}

async function getPrayerTimes(city, fetchText, now = new Date()) {
  const L = londonNow(now);
  const html = await fetchText(buildUrl(city, L.y, L.m));
  const { times, next } = extractToday(html, L);
  return { city, ...times, next };
}

const api = { CITIES, SOURCE_BASE, londonNow, buildUrl, parseRows, extractToday, getPrayerTimes };
if (typeof module !== "undefined" && module.exports) module.exports = api;   // Node / Homey
if (typeof globalThis !== "undefined") globalThis.PrayerCore = api;           // Scriptable
